File: external/lumscript/benchmarks/mandel.js
// Mandelbrot benchmark
// Node.js port of mandel.ls

function norm2(re, im) {
	return re * re - im * (-im);
}

function cabs(re, im) {
	return Math.sqrt(norm2(re, im));
}

function level(x, y) {
	let zre = x;
	let zim = y;

	for (let l = 0; l < 255; l++) {
		const tre = zre * zre - zim * zim;
		const tim = zre * zim + zim * zre;

		zre = tre + x;
		zim = tim + y;

		if (cabs(zre, zim) > 2.0) {
			return l;
		}
	}

	return 255;
}

function main() {
	const xmin = -2.0;
	const ymin = -2.0;
	const n = 256;

	const dx = 4.0 / n;
	const dy = 4.0 / n;

	let result = 0;
	for (let i = 0; i < n; i++) {
		const x = xmin + i * dx;
		for (let j = 0; j < n; j++) {
			const y = ymin + j * dy;
			result += level(x, y);
		}
	}

	return result;
}

const start = performance.now();
const result = main();
const elapsed = performance.now() - start;
console.error(`Time: ${elapsed.toFixed(2)} ms`);
console.log(result);

File: external/lumscript/benchmarks/mandel.c
/* Mandelbrot benchmark
 * C port of mandel.ls
 */

#include <math.h>
#include <stdio.h>
#include <windows.h>

static double now_ms(void) {
	LARGE_INTEGER freq, counter;
	QueryPerformanceFrequency(&freq);
	QueryPerformanceCounter(&counter);
	return (double)counter.QuadPart * 1000.0 / (double)freq.QuadPart;
}

static double norm2(double re, double im) {
	return re * re - im * (-im);
}

static double cabs_(double re, double im) {
	return sqrt(norm2(re, im));
}

static int level(double x, double y) {
	double zre = x;
	double zim = y;

	for (int l = 0; l < 255; l++) {
		const double tre = zre * zre - zim * zim;
		const double tim = zre * zim + zim * zre;

		zre = tre + x;
		zim = tim + y;

		if (cabs_(zre, zim) > 2.0) {
			return l;
		}
	}

	return 255;
}

static int run(void) {
	const double xmin = -2.0;
	const double ymin = -2.0;
	const int n = 256;

	const double dx = 4.0 / n;
	const double dy = 4.0 / n;

	int result = 0;
	for (int i = 0; i < n; i++) {
		const double x = xmin + i * dx;
		for (int j = 0; j < n; j++) {
			const double y = ymin + j * dy;
			result += level(x, y);
		}
	}

	return result;
}

int main(void) {
	const double start = now_ms();
	const int result = run();
	const double elapsed = now_ms() - start;
	fprintf(stderr, "Time: %.2f ms\n", elapsed);
	printf("%d\n", result);
	return 0;
}
